aggregator.go

```diff
@@ -0,0 +1,300 @@
+	RootCid                   cid.Cid
+	UniqueBlockCount          uint64 // optional amount of blocks in dag, recorded in manifest
+	UniqueBlockCumulativeSize uint64 // optional dag size, used as the TSize in the unixfs link entry and recorded in manifest
+}
+
+type nodeMap map[string]*merkledag.ProtoNode
+
+// Aggregate de-duplicates and orders the supplied list of `AggregateDagEntry`-es
+// and adds them into a two-level UnixFSv1 directory structure. The intermediate
+// blocks comprising the directory tree and the manifest json file are written
+// to the supplied DAGService. No "temporary blocks" are produced in the process:
+// everything written to the DAGService is part of the final DAG capped by the
+// final `aggregateRoot`.
+func Aggregate(ds ipldformat.DAGService, toAggregate []AggregateDagEntry) (aggregateRoot cid.Cid, err error) {
+
+	dags := make(map[string]*ManifestDagEntry, len(toAggregate))
+
+	for _, d := range toAggregate {
+
+		cv1 := cid.Undef
+		if d.RootCid.Version() == 1 {
+			cv1 = d.RootCid
+		} else {
+			cv1 = cid.NewCidV1(d.RootCid.Type(), d.RootCid.Hash())
+		}
+
+		// do not consider identity multihashes for aggregation
+		if cv1.Prefix().MhType == uint64(multicodec.Identity) {
+			continue
+		}
+
+		cv0 := cid.Undef
+		if cv1.Prefix().Codec == uint64(multicodec.DagPb) &&
+			cv1.Prefix().MhType == uint64(multicodec.Sha2_256) &&
+			cv1.Prefix().MhLength == 32 {
+			cv0 = cid.NewCidV0(cv1.Hash())
+		}
+
+		dagName := cv1.String()
+
+		if _, exists := dags[dagName]; !exists {
+			dags[dagName] = &ManifestDagEntry{
+				rootCid:    cv1,
+				RecordType: DagAggregateEntry,
+				DagCidV1:   dagName,
+			}
+			if cv0 != cid.Undef {
+				dags[dagName].DagCidV0 = cv0.String()
+			}
+			if d.UniqueBlockCumulativeSize != 0 {
+				v := d.UniqueBlockCumulativeSize
+				dags[dagName].DagSize = &v
+			}
+			if d.UniqueBlockCount != 0 {
+				v := d.UniqueBlockCount
+				dags[dagName].NodeCount = &v
+			}
+		}
+	}
+
+	if len(dags) == 0 {
+		return cid.Undef, xerrors.New("no valid entries to aggregate")
+	}
+
+	// unixfs sorts internally, so we need to pre-sort to match up insertion indices
+	sortedDags := make([]*ManifestDagEntry, 0, len(dags))
+	for _, d := range dags {
+		sortedDags = append(sortedDags, d)
+	}
+	sort.Slice(sortedDags, func(i, j int) bool {
+		return strings.Compare(sortedDags[i].DagCidV1, sortedDags[j].DagCidV1) < 0
+	})
+
+	// innermost layer, 4 bytes off end
+	innerNodes := make(nodeMap)
+	for _, d := range sortedDags {
+
+		parentName := d.DagCidV1[:3] + `...` + d.DagCidV1[len(d.DagCidV1)-4:]
+		if _, exists := innerNodes[parentName]; !exists {
+			innerNodes[parentName] = emptyDir()
+		}
+
+		var dagSize uint64
+		if d.DagSize != nil {
+			dagSize = *d.DagSize
+		}
+
+		if err := innerNodes[parentName].AddRawLink(d.DagCidV1, &ipldformat.Link{
+			Size: dagSize,
+			Cid:  d.rootCid,
+		}); err != nil {
+			return cid.Undef, err
+		}
+		d.PathIndexes[2] = len(innerNodes[parentName].Links()) - 1
```

Comment:
Q: could replace `len(...)` if we dont `_, d : =range` and instead `idx, d := range` and use `idx` to avoid possible alloc in `Links()`?

Reply:
That `range sortedDags` is over *all* dags. Whereas the `len()` is for "the current parent directory". This is why I do not need `Patch()`-like functionality and do not create orphaned blocks at all: everything is built incrementally "as it comes".

Reply:
Agree... I explained quite badly my point.  The point was that since we're iterating in order, we could keep an index of how many links we have added and rest it after every parentName change... but not sure is worth that opt.

Reply:
Ohhhhh I see... Hm... Let me see how much worse would it be read-wise...

Reply:
@jsign wait no - it won't work. The parent changes "randomly" - the parent is selected based on the suffix, but we order the dags based on their full name ( prefix-sort )

Reply:
True. So the `sortedDags` "sorted" criteria is only useful for the manifest creation, not really for the loops here.
(Double checking; we have to eventually sort it anyway...)

Reply:
Nope - you have to do both.
The pigeonholing works based on the suffix.
But within the same directory entries will be alpha-sorted. So we have to sort as well so that the path-indices will match up. I say so much here, though hard to understand without the context: https://github.com/filecoin-project/go-dagaggregator-unixfs/pull/1/files#diff-b4395685fe46f5e5c29b6ef64ea0ce8a79f17a8c37abec7246d3843b9ed740b8R132

Reply:
Oh I see, links should be inserted in order for UnixFS  :+1:

Reply:
More like - go-unixfs will sort *for us*. So if we do not do sort ourselves our `len(innerNodes[parentName].Links())` won't come out correctly
https://github.com/ipfs/go-merkledag/blob/v0.3.2/coding.go#L41